feat!: add C `ndarray` API and refactor `blas/ext/base/sapx`

headlessNode opened the pull request:
## Description

> What is the purpose of this pull request?

This pull request:

-   add C `ndarray` API and refactor `blase/ext/base/sapx`

## Related Issues

> Does this pull request have any related issues?

No. 

## Questions

> Any questions for reviewers of this pull request?

No.

## Other

> Any other information relevant to this pull request? This may include screenshots, references, and/or implementation notes.

No.

## Checklist

> Please ensure the following tasks are completed before submitting this pull request.

-   [x] Read, understood, and followed the [contributing guidelines][contributing].

* * *

@stdlib-js/reviewers

[contributing]: https://github.com/stdlib-js/stdlib/blob/develop/CONTRIBUTING.md

headlessNode committed:
feat: add C ndarray API and refactor

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: passed
  - task: lint_package_json
    status: passed
  - task: lint_repl_help
    status: passed
  - task: lint_javascript_src
    status: passed
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: passed
  - task: lint_javascript_tests
    status: na
  - task: lint_javascript_benchmarks
    status: passed
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: missing_dependencies
  - task: lint_c_examples
    status: missing_dependencies
  - task: lint_c_benchmarks
    status: missing_dependencies
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: passed
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

---
type: pre_push_report
description: Results of running various checks prior to pushing changes.
report:
  - task: run_javascript_examples
    status: na
  - task: run_c_examples
    status: na
  - task: run_cpp_examples
    status: na
  - task: run_javascript_readme_examples
    status: na
  - task: run_c_benchmarks
    status: na
  - task: run_cpp_benchmarks
    status: na
  - task: run_fortran_benchmarks
    status: na
  - task: run_javascript_benchmarks
    status: na
  - task: run_julia_benchmarks
    status: na
  - task: run_python_benchmarks
    status: na
  - task: run_r_benchmarks
    status: na
  - task: run_javascript_tests
    status: na
---

headlessNode commented on lib/node_modules/@stdlib/blas/ext/base/sapx/README.md:
````diff
@@ -92,17 +92,17 @@ sapx.ndarray( x.length, 5.0, x, 1, 0 );
+While [`typed array`][mdn-typed-array] views mandate a view offset based on the underlying buffer, the offset parameter supports indexing semantics based on a starting index. For example, to access only the last three elements of the strided array:
 
 ```javascript
 var Float32Array = require( '@stdlib/array/float32' );
 
 var x = new Float32Array( [ 1.0, -2.0, 3.0, -4.0, 5.0, -6.0 ] );
 
 sapx.ndarray( 3, 5.0, x, 1, x.length-3 );
-// x => <Float32Array>[ 1.0, -2.0, 3.0, 1.0, 10.0, -1.0 ]
+// x => <Float32Array>[ 6.0, 3.0, 8.0, -4.0, 5.0, -6.0 ]
 ```
````
@kgryte Here I was getting a lint error which prevented me from pushing. I made this change to be able to push the commit.

I tried finding the cause but couldn't find any. The tests are passing, I also checked by changing the example in `examples/index.js`, the `offset` parameter worked there correctly. But for some reason here it is not working as expected. Any insights?

EDIT: The CI failure errors are the opposite of the errors I was getting in Local

headlessNode commented on lib/node_modules/@stdlib/blas/ext/base/sapx/lib/ndarray.native.js:
```diff
@@ -20,37 +20,31 @@
-* @param {number} alpha - scalar
+* @param {number} alpha - scalar constant
 * @param {Float32Array} x - input array
-* @param {integer} stride - index increment
-* @param {NonNegativeInteger} offset - starting index
+* @param {integer} strideX - stride length
+* @param {NonNegativeInteger} offsetX - starting index
 * @returns {Float32Array} input array
 *
 * @example
 * var Float32Array = require( '@stdlib/array/float32' );
 *
 * var x = new Float32Array( [ 1.0, -2.0, 3.0, -4.0, 5.0, -6.0 ] );
-* var alpha = 5.0;
 *
-* sapx( 3, alpha, x, 1, x.length-3 );
-* // x => <Float32Array>[ 1.0, -2.0, 3.0, 1.0, 10.0, -1.0 ]
+* sapx( 3, 5.0, x, 1, x.length-3 );
+* // x => <Float32Array>[ 6.0, 3.0, 8.0, -4.0, 5.0, -6.0 ]
 */
```
same comment here. Notice that the doctest for `ndarray.js` passses but for `ndarray.native.js` it fails.

headlessNode commented on lib/node_modules/@stdlib/blas/ext/base/sapx/README.md:
```diff
@@ -92,17 +92,17 @@ sapx.ndarray( x.length, 5.0, x, 1, 0 );
 
 sapx.ndarray( 3, 5.0, x, 1, x.length-3 );
-// x => <Float32Array>[ 1.0, -2.0, 3.0, 1.0, 10.0, -1.0 ]
+// x => <Float32Array>[ 6.0, 3.0, 8.0, -4.0, 5.0, -6.0 ]
```
```suggestion
// x => <Float32Array>[ 1.0, -2.0, 3.0, 1.0, 10.0, -1.0 ]
```

headlessNode commented on lib/node_modules/@stdlib/blas/ext/base/sapx/lib/ndarray.native.js:
```diff
@@ -20,37 +20,31 @@
-* sapx( 3, alpha, x, 1, x.length-3 );
-* // x => <Float32Array>[ 1.0, -2.0, 3.0, 1.0, 10.0, -1.0 ]
+* sapx( 3, 5.0, x, 1, x.length-3 );
+* // x => <Float32Array>[ 6.0, 3.0, 8.0, -4.0, 5.0, -6.0 ]
```
```suggestion
* // x => <Float32Array>[ 1.0, -2.0, 3.0, 1.0, 10.0, -1.0 ]
```

headlessNode committed:
docs: apply review suggestion

Signed-off-by: Muhammad Haris <101793258+headlessNode@users.noreply.github.com>

headlessNode committed:
docs: apply review suggestion

Signed-off-by: Muhammad Haris <101793258+headlessNode@users.noreply.github.com>

headlessNode commented on lib/node_modules/@stdlib/blas/ext/base/sapx/examples/c/example.c:
```diff
@@ -21,16 +21,16 @@
 
-	// Add a constant to each element:
-	c_sapx( N, 5.0f, x, strideX );
+	// Fill the array:
```
```suggestion
	// Add a constant to each element:
```

headlessNode committed:
docs: apply review suggestion

Signed-off-by: Muhammad Haris <101793258+headlessNode@users.noreply.github.com>

kgryte commented on lib/node_modules/@stdlib/blas/ext/base/sapx/src/main.c:
```diff
@@ -0,0 +1,84 @@
+		// If we have a remainder, run a clean-up loop...
+		if ( m > 0 ) {
+			for ( i = 0; i < m; i++ ) {
+				X[ i ] += alpha;
```
```suggestion
				X[ ix ] += alpha;
```

This is your bug and the reason for the lint errors.

headlessNode committed:
fix: apply code review suggestion

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: na
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: na
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: na
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: missing_dependencies
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: na
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

---
type: pre_push_report
description: Results of running various checks prior to pushing changes.
report:
  - task: run_javascript_examples
    status: na
  - task: run_c_examples
    status: na
  - task: run_cpp_examples
    status: na
  - task: run_javascript_readme_examples
    status: na
  - task: run_c_benchmarks
    status: na
  - task: run_cpp_benchmarks
    status: na
  - task: run_fortran_benchmarks
    status: na
  - task: run_javascript_benchmarks
    status: na
  - task: run_julia_benchmarks
    status: na
  - task: run_python_benchmarks
    status: na
  - task: run_r_benchmarks
    status: na
  - task: run_javascript_tests
    status: na
---

headlessNode committed:
docs: wrap repl line

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: na
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: passed
  - task: lint_javascript_src
    status: na
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: na
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: na
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

---
type: pre_push_report
description: Results of running various checks prior to pushing changes.
report:
  - task: run_javascript_examples
    status: na
  - task: run_c_examples
    status: na
  - task: run_cpp_examples
    status: na
  - task: run_javascript_readme_examples
    status: na
  - task: run_c_benchmarks
    status: na
  - task: run_cpp_benchmarks
    status: na
  - task: run_fortran_benchmarks
    status: na
  - task: run_javascript_benchmarks
    status: na
  - task: run_julia_benchmarks
    status: na
  - task: run_python_benchmarks
    status: na
  - task: run_r_benchmarks
    status: na
  - task: run_javascript_tests
    status: na
---

kgryte commented on lib/node_modules/@stdlib/blas/ext/base/sapx/README.md:
```diff
@@ -141,6 +142,125 @@ console.log( x );
+    // Specify a stride:
+    const int strideX = 1;
+
+    // Fill the array:
```
```suggestion
    // Add a constant to each element:
```

kgryte committed:
docs: update comment

Signed-off-by: Athan <kgryte@gmail.com>

kgryte reviewed (approved):
LGTM
